text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,590 @@
+| `receive`      | One or more messages are requested by a consumer. |
+| `deliver`      | One or more messages are passed to a consumer. |
+| `settle`       | One or more messages are settled. |
+
+For further details about each of those operations refer to the [section about trace structure](#trace-structure).
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+
+The kind of `settle` spans should be set according to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind),
+e. g. it should be to `CLIENT` if the `settle` spans models a synchronous call
+to the intermediary.
+
+Setting span kinds according to this table ensures, that span links between
+consumers and producers always go from a `PRODUCER` span on the producer side to
+a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
+linked traces without the need for additional semantic hints.
+
+### Span relationships
+
+#### Producer spans
+
+"Publish" spans SHOULD be created for operations of providing messages for
+sending or publishing to an intermediary. A single "Publish" span can account
+for a single message, or for multiple messages (in the case of providing
+messages in batches). "Create" spans MAY be created. A single "Create" span
+SHOULD account only for a single message.
+
+If a "Create" span exists for a message, its context SHOULD be attached to
+the message. If no "Create" span exists, the context of the related "Publish"
+span SHOULD be attached to the message.
+
+#### Consumer spans
+
+##### Push-based scenarios
+
+"Deliver" spans SHOULD be created for operations of passing messages to the
+application when those operations are not initiated by the application
+code.
+
+##### Pull-based scenarios
+
+"Receive" spans SHOULD be created for operations of passing messages to the
+application when those operations are initiated by the application code.
+
+##### General considerations
+
+"Deliver" or "Receive" spans MUST NOT be created for messages which are not
+forwarded to the caller, but are pre-fetched or cached by messaging
+libraries or SDKs.
+
+A single "Deliver" or "Receive" span can account for a single message, for
+multiple messages (in case messages are passed for processing as batches), or
+for no message at all (if it is signalled that no messages were received).  For
+each message it accounts for, the "Deliver" or "Receive" span SHOULD link to
+the "Create" or "Publish" span for the message. In addition, if it is possible
```

Comment:
Should we phrase this part in a way that it "favors" or give "precedence" for adding it as a parent when possible, and Links being the "fallback" if not? Because (IIRC) that's what Tyler was advocating for and I think it's what a lot of users would prefer. 

In my interpretation the way it is now, it seems the link option takes precedence (or it is more important?) than the second. 

Also: Should we spec something like: If added as parent, then don't add links? Or do we always want the links there for consistency? Thinking as a user consuming a trace, wouldn't it be awkward that it's all in the same trace, and there's a link to "itself"?

Reply:
> In my interpretation the way it is now, it seems the link option takes precedence (or it is more important?) than the second.

Yes, it's my intention to convey this. In my opinion, the link should always be there. This gives consistency across all scenarios, it's something backends can rely on, and it allows to integrate intermediary instrumentation.

Reply:
I agree it makes things more consistent if we always add links. But I think people consuming these conventions would benefit of having a clear message that parents are still the "desired outcome" _when_ it's possible to do so.

Reply:
I will mark this as resolved, as we already say here:

>In addition, if it is possible the "Create" or "Publish" span MAY be set as a parent of the "Deliver" or "Receive" span.